File: MergeSort.py
def MergeSort(arr,p,r):
    if p>=r:
        return
    q=(p+r)//2
    MergeSort(arr,p,q)
    MergeSort(arr,q+1,r)
    Merge(arr,p,q,r)

def Merge(arr,p,q,r):
    n1=q-p+1
    n2=r-q
    l=[]
    r=[]
    for i in range(n1):
        l[i]=arr[p+i]
    for i in range(n2):
        r[i]=arr[q+i+1]
    l[n1+1]=10e9
    r[n2+1]=10e9
    i=1
    j=1
    for k in range(p,r):
        if l[i]<=r[j]:
            arr[k]=l[i]
            i+=1
        else:
            arr[k]=r[j]
            j+=1

arr=[2,5,8,3,9,0,1,5,7]
MergeSort(arr,0,8)
print(arr)

File: mergesort.cpp
# include<iostream>
# include<time.h>
# include<climits>
# include<chrono>
using namespace std;


void merging(int *arr,int start,int mid,int last){
    int i=start,j=mid+1,k=0;
    int *temp= new int[last-start+1];
    while(i<=mid && j<=last){
        if(*(arr+i)<= *(arr+j)){
        temp[k]=arr[i]; k++; i++;

        }
        else {temp[k]=arr[j]; k++; j++;}
        }
    while(i<=mid){
        temp[k]=arr[i]; k++; i++;
        }
    while(j<=last){
        temp[k]=arr[j]; k++; j++;
        }
    i=start;
    k=0;
    while(i<=last){
        arr[i]=temp[k];
        i++;
        k++;
        }
}

void mergesort(int *arr,int start , int endidx){
    if(start>=endidx) return;
    int mid=(start+endidx)/2;
    mergesort(arr,start,mid);
    mergesort(arr,mid+1,endidx);
    merging(arr,start,mid,endidx);

}

void mergeSort(int input[], int siz){

    mergesort(input,0,siz-1);

}


void swa(int &a, int &b){
    int temp=a; a=b;b=temp;
}







int main(){
    int n;
    int t;
    cin>>t;
    for(int i=0; i<t ;i++){
    cin>>n;
    int * arr= new int [n];
    for(int i=0; i<n; i++){
        arr[i]=rand();
    }
    auto start = std::chrono::high_resolution_clock::now();
    mergeSort(arr,n);


    auto elapsed = std::chrono::high_resolution_clock::now() - start;
     long long microseconds = std::chrono::duration_cast<std::chrono::microseconds>(elapsed).count();
    cout<<"time:"<<microseconds<<"microseconds \n";
    //cout<<"Array: \n";
   /* for(int i=0; i<n; i++){
        cout<<arr[i]<<" ";
    }
    cout<<"\n";
*/



}
}
